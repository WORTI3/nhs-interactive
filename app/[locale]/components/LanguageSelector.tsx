import { Fragment } from 'react'
import { Menu, Transition } from '@headlessui/react'
import { IconWorld } from '@tabler/icons-react';
import Link from "next-intl/link";
import { LOCALES } from '../../utils/constants';

function classNames(...classes: any[]) {
  return classes.filter(Boolean).join(' ')
}

export default function LanguageSelector({translate, locale}: {translate: any, locale: string}) {
  return (
    <Menu as='div' className='relative max-lg:absolute max-lg:right-4 float-right inline-block text-left'>
      <div>
        {/* Hover style made need changed to match NHS */}
        <Menu.Button id='language-button' className='inline-flex w-full justify-center gap-x-1.5 rounded-md bg-white px-3 py-2 text-sm font-semibold text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 hover:bg-gray-50'>
          {translate('button.language')}
          <IconWorld
            className='-mr-1 h-5 w-5 text-gray-400'
            aria-hidden='true'
          />
        </Menu.Button>
      </div>

      <Transition
        as={Fragment}
        enter='transition ease-out duration-100'
        enterFrom='transform opacity-0 scale-95'
        enterTo='transform opacity-100 scale-100'
        leave='transition ease-in duration-75'
        leaveFrom='transform opacity-100 scale-100'
        leaveTo='transform opacity-0 scale-95'
      >
        <Menu.Items className='absolute right-0 z-10 mt-2 w-56 origin-top-right rounded-md bg-white shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none'>
          <div className='langs'>
            {LOCALES.map(lang => {
              if (lang !== locale) {
                return (
                  <Menu.Item key={lang} data-testid="language-item">
                    {({ active }) => (
                      <Link
                        href='/'
                        locale={lang}
                        className={classNames(
                          active ? "bg-gray-100 text-gray-900" : "text-gray-700",
                          "block px-4 py-2 text-sm rounded-md"
                        )}
                      >
                        <p className='mb-0 py-2'>
                          <span
                            className={`mr-4 h-5 w-5 rounded-md fi fis fi-${lang === 'en' ? 'gb-eng' : lang === 'cy' ? 'gb-wls' : 'gb-sct'}`}
                            aria-hidden='true'
                          />
                          {lang === 'en' ? 'English' : lang === 'cy' ? 'Cymraeg' : 'Gaelic'}
                        </p>
                      </Link>
                    )}
                  </Menu.Item>
                );
              }
              return null;
            })}
          </div>
        </Menu.Items>
      </Transition>
    </Menu>
  );
}
