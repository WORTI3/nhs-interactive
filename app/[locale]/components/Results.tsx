import { FC } from "react";
import { UNRELATED_ANSWER } from "../../utils/constants";
import { ResultProps } from "../../utils/interfaces";
import { useApiLog } from "../../utils/hooks";
import { LogData } from "@/types";
import { useTranslations } from "next-intl";

export const Results: FC<ResultProps> = ({ searchQuery, answer, done, onReset }) => {
  const translate = useTranslations('results');
  const logData: LogData = { searchQuery, answer };
  useApiLog(logData, done);
  return (
    <>
      <p className='text-md text-nhs-blue mb-2'>{translate('question')}</p>
      <p className='italic'>{searchQuery.query} (?)</p>
      <hr className='nhsuk-section-break nhsuk-section-break--m nhsuk-section-break--visible' />
      { answer === UNRELATED_ANSWER ? <>
        <p>{translate('error.paragraphOne')}</p>
        <div className='nhsuk-details__text mb-6'>
          <p>{translate('error.paragraphTwo')}</p>
          <ul>
            <li>{translate('error.list.one')}</li>
            <li>{translate('error.list.two')}</li>
            <li>{translate('error.list.three')}</li>
          </ul>
        </div>
        </> : <>
          <p className='text-nhs-blue mb-2'>{translate('answer')}</p>
          <p className='overflow-auto'>
            {answer}
          </p>
        </>
      }
      
      {(done && answer !== UNRELATED_ANSWER) && (
        <>
          <hr className='nhsuk-section-break nhsuk-section-break--m nhsuk-section-break--visible' />
          <p className='text-nhs-blue mb-2'>{translate('sources')}</p>
          {searchQuery.sourceLinks.map((source, index) => (
            <div key={index} className='my-2 overflow-auto max-sm:flex'>
              <a
                className='hover:cursor-pointer hover:underline'
                target='_blank'
                rel='noopener noreferrer'
                href={source}
              >
                <span className="max-sm:flex-1  ">
                  {searchQuery.sourceHeadings[index] ?? 'NHS Search'}
                  <span className="ml-4 max-sm:float-none sm:float-right inline-flex items-center rounded-md bg-sky-50 max-sm:px-1.5 px-2 py-1 max-sm:text-xs text-xs max-sm:font-semibold font-bold text-sky-600 ring-1 ring-inset ring-sky-600/90 hover:bg-sky-100">
                    Verified source: {source.split("//")[1].split("/")[0].replace("www.", "")}
                  </span>
                </span>
              </a>
            </div>
          ))}
          <hr className='nhsuk-section-break nhsuk-section-break--m nhsuk-section-break--visible' />
        </>
      )}
      { (done || answer === UNRELATED_ANSWER) && (
        <button className="nhsuk-button" onClick={onReset}>{translate('button.newQuery')}</button>
      )}
    </>
  );
};
