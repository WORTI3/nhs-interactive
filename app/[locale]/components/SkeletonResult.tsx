export const SkeletonResult = () => {
  return (
    <>
    <div className="animate-pulse space-y-6">
      <div className="h-6 w-[88px] bg-gray-200 rounded"></div>
      <div className="flex justify-between space-x-4">
        <div className="h-6 bg-gray-200 rounded w-1/2"></div>
        <div className="h-6 bg-gray-200 rounded w-[68px]"></div>
      </div>
    </div>
    <div className="animate-pulse space-y-6">
      <div className="h-1 bg-gray-200 rounded mb-10 mt-3"></div>
      <div className="space-y-4">
        <div className="h-4 bg-gray-200 rounded"></div>
        <div className="h-4 bg-gray-200 rounded w-5/6"></div>
        <div className="h-4 bg-gray-200 rounded w-5/6"></div>
        <div className="h-4 bg-gray-200 rounded w-4/6"></div>
        <div className="h-4 bg-gray-200 rounded w-1/2"></div>
      </div>
      <div className="h-14 bg-gray-200 rounded w-[196px]"></div>
    </div>
    </>
  );
};
