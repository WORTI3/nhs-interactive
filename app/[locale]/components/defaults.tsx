import { OpenAIModel } from "../../utils/constants";

/**
 * `export const ModelInfo` is defining a React functional component that returns a paragraph element displaying
 * information about a specific OpenAI model. The model information is accessed through the `OpenAIModel` constant defined
 * in the `constants.ts` file. This component can be imported and used in other parts of the codebase.
 * 
 * @function
 * @name ModelInfo
 * @kind variable
 * @returns {JSX.Element}
 * @exports
 */
export const ModelInfo = () => {
  return (
    <p className="nhsuk-body-s nhsuk-u-secondary-text-color nhsuk-u-margin-top-7 nhsuk-u-margin-bottom-0">
      Model: { OpenAIModel.DAVINCI_TURBO }
    </p>
  );
}

/**
 * This code exports a function named `makeSourcesLinks` that takes in two parameters: `answer` and `sourceLinks`. The
 * function splits the `answer` string using a regular expression and maps over the resulting array. If a part of the
 * string matches the regular expression `/\[[0-9]+\]/`, it extracts the number within the square brackets, uses it to
 * retrieve a link from the `sourceLinks` array, and returns an anchor tag with the link and the original part of the
 * string. If the part of the string does not match the regular expression, it returns the original part of the string. The
 * function then returns an array of elements that can be rendered in the UI.
 * 
 * @function
 * @name makeSourcesLinks
 * @kind variable
 * @exports
 */
export const makeSourcesLinks = (answer: string, sourceLinks: string[]) => {
  const elements = answer.split(/(\[[0-9]+\])/).map((part, index) => {
    if (/\[[0-9]+\]/.test(part)) {
      const link = sourceLinks[parseInt(part.replace(/[\[\]]/g, "")) - 1];

      return (
        <a
          key={index}
          className="hover:cursor-pointer text-nhs-blue text-base"
          href={link}
          target="_blank"
          rel="noopener noreferrer"
        >
          {part}
        </a>
      );
    } else {
      return part;
    }
  });

  return elements;
};