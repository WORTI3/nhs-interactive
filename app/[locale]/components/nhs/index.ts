export { default as Button } from './Button';
export { default as Input } from './Input';
export { Header } from './Header';
export { Footer } from './Footer';
export { wrapper as Wrapper } from './wrapper';
export * from './WarningCallout';