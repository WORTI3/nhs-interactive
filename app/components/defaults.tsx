
import { IconBrandGithub } from "@tabler/icons-react";
import { OpenAIModel } from "../lib/utils/constants";

/**
 * `export const GitHubIcon = () => {` is defining a React functional component that returns an anchor tag with the GitHub
 * icon from the `@tabler/icons-react` library. The anchor tag has a link to the GitHub profile of the user `worti3`. This
 * component can be imported and used in other parts of the codebase.
 * 
 * @function
 * @name GitHubIcon
 * @kind variable
 * @returns {JSX.Element}
 * @exports
 */
export const GitHubIcon = () => {
  return (
    <a
      className="absolute top-2 right-2 p-4 cursor-pointer text-white hover:text-yellow-400 visited:text-white"
      href="https://github.com/worti3"
      target="_blank"
      rel="noreferrer"
      aria-label="Github Profile"
    >
      <IconBrandGithub />
    </a>
  );
};

/**
 * `export const ModelInfo` is defining a React functional component that returns a paragraph element displaying
 * information about a specific OpenAI model. The model information is accessed through the `OpenAIModel` constant defined
 * in the `constants.ts` file. This component can be imported and used in other parts of the codebase.
 * 
 * @function
 * @name ModelInfo
 * @kind variable
 * @returns {JSX.Element}
 * @exports
 */
export const ModelInfo = () => {
  return (
    <p className="nhsuk-body-s nhsuk-u-secondary-text-color nhsuk-u-margin-top-7 nhsuk-u-margin-bottom-0">
      Model: { OpenAIModel.DAVINCI_TURBO }
    </p>
  );
}

/**
 * This code exports a function named `makeSourcesLinks` that takes in two parameters: `answer` and `sourceLinks`. The
 * function splits the `answer` string using a regular expression and maps over the resulting array. If a part of the
 * string matches the regular expression `/\[[0-9]+\]/`, it extracts the number within the square brackets, uses it to
 * retrieve a link from the `sourceLinks` array, and returns an anchor tag with the link and the original part of the
 * string. If the part of the string does not match the regular expression, it returns the original part of the string. The
 * function then returns an array of elements that can be rendered in the UI.
 * 
 * @function
 * @name makeSourcesLinks
 * @kind variable
 * @exports
 */
export const makeSourcesLinks = (answer: string, sourceLinks: string[]) => {
  const elements = answer.split(/(\[[0-9]+\])/).map((part, index) => {
    if (/\[[0-9]+\]/.test(part)) {
      const link = sourceLinks[parseInt(part.replace(/[\[\]]/g, "")) - 1];

      return (
        <a
          key={index}
          className="hover:cursor-pointer text-nhs-blue text-base"
          href={link}
          target="_blank"
          rel="noopener noreferrer"
        >
          {part}
        </a>
      );
    } else {
      return part;
    }
  });

  return elements;
};