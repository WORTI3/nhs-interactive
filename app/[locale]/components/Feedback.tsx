import { Dialog, DialogTitle } from '@headlessui/react';
import { useState } from 'react';
import { Button } from './nhs';
import { useTranslations } from 'next-intl';

const FeedbackDialog = ({ isOpen, onClose, onSubmit }: {isOpen: any, onClose: any, onSubmit: any}) => {
  const translate = useTranslations('search');
  const [comment, setComment] = useState('');
  const [helpful, setHelpful] = useState(false);
  const [error, setError] = useState<string>("");

  const handleSubmit = () => {
    onSubmit({ comment, helpful });
    onClose();
  };

  return (
    <Dialog open={isOpen} onClose={onClose} className="fixed inset-0 overflow-hidden">
      <div className="absolute inset-0 overflow-hidden">
        <div className="absolute inset-0 bg-black opacity-50" />
        <div className="fixed inset-y-0 right-0 max-w-full flex">
          <div className="w-screen max-w-md">
            <div className="h-full bg-white p-4 shadow-xl">
              <DialogTitle>Provide Feedback</DialogTitle>
              <textarea
                className="w-full border rounded mt-4 p-2"
                placeholder="Leave a note..."
                value={comment}
                onChange={(e) => setComment(e.target.value)}
              />
              <div className="mt-4">
                <label>
                  <input
                    type="checkbox"
                    checked={helpful}
                    onChange={(e) => setHelpful(e.target.checked)}
                  />
                  Was this helpful?
                </label>
              </div>
              <div className="mt-4">
                <button
                  className="bg-blue-500 text-white p-2 rounded"
                  onClick={handleSubmit}
                >
                  Submit
                </button>
                <Button
                  id='submit'
                  as='a'
                  data-prevent-double-click
                  onClick={handleSubmit}
                  disabled={!!error && comment === ''}
                >
                  {translate('button.submit')}
                </Button>
                <button
                  className="ml-2 p-2"
                  onClick={onClose}
                >
                  Cancel
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </Dialog>
  );
};

export default FeedbackDialog;
