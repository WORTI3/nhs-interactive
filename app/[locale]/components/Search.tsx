"use client";

import { ChangeEvent, FC, KeyboardEvent, useState } from "react";
import { useTranslations } from "next-intl";
import { z } from "zod";
import { Button, Input } from "./nhs";
import { ALLOWED_SEARCH_CHARS_REGEX, SearchProps, fetchAnswer } from "../../utils";

const querySchema = z.string().min(3).max(100).regex(ALLOWED_SEARCH_CHARS_REGEX);

export const Search: FC<SearchProps> = ({
  onSearch,
  onResultUpdate: onAnswerUpdate,
  onDone,
}) => {
  const translate = useTranslations('search');
  const [query, setQuery] = useState<string>("");
  const [loading, setLoading] = useState<boolean>(false);
  const [error, setError] = useState<string>("");
  
  const handleKeyDown = (e: KeyboardEvent<HTMLInputElement>) => {
    if (e.key === "Enter") {
      handleSearch();
    }
  };

  const handleSearch = async () => {
    const result = querySchema.safeParse(query);
    if (!result.success) {
      // dont care for multiple errors here so using the first in array is fine.
      setError(translate(`errors.${result.error.issues[0].code}`));
      return;
    }

    setLoading(true);
    try {
      await fetchAnswer(query, onAnswerUpdate, onSearch, onDone, translate('lang'));
    } catch (error) {
      setError('Service currently unavailable. Please try again later.');
    }
    setLoading(false);
  };

  return (
    <>
      {/* todo: skeleton loading here? */}
      {loading ? (
        <>
          <span className='inline-flex'>
            <div data-testid='animated-progress' className='h-6 w-6 animate-spin rounded-full border-2 border-solid border-current border-r-transparent align-[-0.125em] motion-reduce:animate-[spin_1.5s_linear_infinite] mr-2'></div>
            <p>{translate('loading')}</p>
          </span>
        </>
      ) : (
        <>
        <div className='relative w-full'>
          <Input
            id="search-input"
            label={translate('query')}
            name="search"
            hint={translate('hint')}
            placeholder={translate('placeholder')}
            value={query}
            type="text"
            onChange={(e: ChangeEvent<HTMLInputElement>) => setQuery(e.target.value)}
            onKeyDown={handleKeyDown}
            error={error}
          />
          <Button
            id='submit'
            as='a'
            data-prevent-double-click
            onClick={handleSearch}
            disabled={!!error && query === ''}
          >
            {translate('button.submit')}
          </Button>
        </div>
        </>
      )}
    </>
  );
};
