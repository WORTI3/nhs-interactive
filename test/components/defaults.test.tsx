import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { GitHubIcon, ModelInfo, makeSourcesLinks } from '../../app/components/defaults';
import { load } from 'cheerio';

describe('GitHubIcon', () => {
  test('rendered component should match snapshot', () => {
    const { container } = render(GitHubIcon());
    expect(container).toMatchSnapshot();
  });

  test('should render the component with the correct links & elements', () => {
    const { container } = render(GitHubIcon());
    const html = container.innerHTML;
    const $ = load(html);  

    expect($('a').length).toBe(1);
    expect($('svg').length).toBe(1);
    expect($('p').length).toBe(0);
  });
});

describe('ModelInfo', () => {
  test('rendered component should match snapshot', () => {
    const { container } = render(ModelInfo());
    expect(container).toMatchSnapshot();
  });

  test('should render the component with the correct links & elements', () => {
    const { container } = render(ModelInfo());
    const html = container.innerHTML;
    const $ = load(html);  

    expect($('a').length).toBe(0);
    expect($('svg').length).toBe(0);
    expect($('p').length).toBe(1);
    expect($('p').text()).toBe("Model: gpt-3.5-turbo");
  });
});

describe('makeSourcesLinks', () => {
  const answer = "The running fox jumped over the hill. They jump so high due to their ability to contract muscles [1] at lightning speed. [2]";
  test('rendered component should match snapshot', () => {
    const { container } = render(<>{ makeSourcesLinks(answer, ["link1", "link2"]) }</>);
    expect(container).toMatchSnapshot();
  });

  test('should render the component with the correct links & elements', () => {
    const { container } = render(<>{ makeSourcesLinks(answer, ["link1", "link2"]) }</>);

    const html = container.innerHTML;
    const $ = load(html);  

    expect($('a').length).toBe(2);
    expect($('a').text()).toBe("[1][2]");
    expect($('a').hasClass('text-blue-500')).toBeTruthy();
    expect($('a').hasClass('hover:cursor-pointer')).toBeTruthy();
    expect($('svg').length).toBe(0);
    expect($('p').length).toBe(0);
  });
});
