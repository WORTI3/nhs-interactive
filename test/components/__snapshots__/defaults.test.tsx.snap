// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitHubIcon rendered component should match snapshot 1`] = `
<div>
  <a
    class="absolute top-2 right-2 p-4 cursor-pointer text-white hover:text-yellow-400 visited:text-white"
    href="https://github.com/worti3"
    rel="noreferrer"
    target="_blank"
  >
    <svg
      class="tabler-icon tabler-icon-brand-github"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9 19c-4.3 1.4 -4.3 -2.5 -6 -3m12 5v-3.5c0 -1 .1 -1.4 -.5 -2c2.8 -.3 5.5 -1.4 5.5 -6a4.6 4.6 0 0 0 -1.3 -3.2a4.2 4.2 0 0 0 -.1 -3.2s-1.1 -.3 -3.5 1.3a12.3 12.3 0 0 0 -6.2 0c-2.4 -1.6 -3.5 -1.3 -3.5 -1.3a4.2 4.2 0 0 0 -.1 3.2a4.6 4.6 0 0 0 -1.3 3.2c0 4.6 2.7 5.7 5.5 6c-.6 .6 -.6 1.2 -.5 2v3.5"
      />
    </svg>
  </a>
</div>
`;

exports[`ModelInfo rendered component should match snapshot 1`] = `
<div>
  <p
    class="nhsuk-body-s nhsuk-u-secondary-text-color nhsuk-u-margin-top-7 nhsuk-u-margin-bottom-0"
  >
    Model: 
    gpt-3.5-turbo
  </p>
</div>
`;

exports[`makeSourcesLinks rendered component should match snapshot 1`] = `
<div>
  The running fox jumped over the hill. They jump so high due to their ability to contract muscles 
  <a
    class="hover:cursor-pointer text-blue-500"
    href="link1"
    rel="noopener noreferrer"
    target="_blank"
  >
    [1]
  </a>
   at lightning speed. 
  <a
    class="hover:cursor-pointer text-blue-500"
    href="link2"
    rel="noopener noreferrer"
    target="_blank"
  >
    [2]
  </a>
</div>
`;
